tutorials/testing-coreutils.md

```diff
@@ -61,108 +61,113 @@ src$ rm -f *.gcda # Get rid of any stale gcov files
 
-File '../../lib/openat.h'
-Lines executed:0.00% of 8
-../../lib/openat.h:creating 'openat.h.gcov'
+## Step 2: Install WLLVM
 
-File '../../src/echo.c'
-Lines executed:18.81% of 101
-../../src/echo.c:creating 'echo.c.gcov'
+One of the difficult parts of testing real software using KLEE is that it must be first compiled so that the final program is an LLVM bitcode file and not a linked executable. The software's build system may be set up to use tools such as 'ar', 'libtool', and 'ld', which do not in general understand LLVM bitcode files.
+
+For coreutils, we use [whole-program-llvm](https://github.com/travitch/whole-program-llvm) (WLLVM), which provides tools for building whole-program LLVM bitcode files from an unmodified C or C++ source package. WLLVM includes four python executables: `wllvm` for C code and `wllvm++` for C++ code, the tool `extract-bc` for extracting the bitcode from a build product (either object file, executable, library, or archive), and the sanity checker `wllvm-sanity-checker` for detecting configuration oversights.
+
+To install _whole-program-llvm_ the easiest way is to use `pip`:
+
+{% highlight bash %}
+$ pip install --upgrade wllvm
 {% endhighlight %}
 
-By default `gcov` will show the number of lines executed in the program (the `.h` files include code which was compiled into `echo.c`).
+To successfully execute WLLVM it is necessary to set the environment variable `LLVM_COMPILER` to the underlying LLVM compiler (either `dragonegg` or `clang`). In this tutorial, we use `clang`:
 
-## Step 2: Build Coreutils with LLVM
+{% highlight bash %}
+$ export LLVM_COMPILER=clang
+{% endhighlight %}
 
-One of the difficult parts of testing real software using KLEE is that it must be first compiled so that the final program is an LLVM bitcode file and not a linked executable. The software's build system may be set up to use tools such as 'ar', 'libtool', and 'ld', which do not in general understand LLVM bitcode files.
+To make the environment variable persistent, add the export to your shell profile (e.g. `.bashrc`).
 
-It depends on the actual project what the best way to do this is. For coreutils, we use a helper script `klee-gcc`, which acts like `llvm-gcc` but adds additional arguments to cause it to emit LLVM bitcode files and to call `llvm-ld` to link executables. This is **not** a general solution, and your mileage may vary.
+## Step 3: Build Coreutils with LLVM
 
 As before, we will build in a separate directory so we can easily access both the native executables and the LLVM versions. Here are the steps:
 
 {% highlight bash %}
 coreutils-6.11$ mkdir obj-llvm
 coreutils-6.11$ cd obj-llvm
-obj-llvm$ ../configure --disable-nls CFLAGS="-g"
+obj-llvm$ CC=wllvm ../configure --disable-nls CFLAGS="-g"
 ... verify that configure worked ...
-obj-llvm$ make CC=/full/path/to/klee/scripts/klee-gcc
-obj-llvm$ make -C src arch hostname CC=/full/path/to/klee/scripts/klee-gcc
+obj-llvm$ CC=wllvm make
```

Comment:
If you configured with `CC` set to `wllvm` it shouldn't be necessary to pass it to `make` as well assuming that coreutils uses autoconf. You'll need to check this though.